src/Hosting/Hosting/src/Internal/HostingApplication.cs

```diff
@@ -29,40 +30,132 @@ public class HostingApplication : IHttpApplication<HostingApplication.Context>
         }
 
         // Set up the request
-        public Context CreateContext(IFeatureCollection contextFeatures)
+        public HostContext CreateContext(IFeatureCollection contextFeatures)
         {
-            var context = new Context();
             var httpContext = _httpContextFactory.Create(contextFeatures);
 
-            _diagnostics.BeginRequest(httpContext, ref context);
+            HostContext hostContext;
+            if (contextFeatures is IHostContextContainer<HostContext> container)
+            {
+                // Initalize the wrapper struct in-place; so its object reference gets set
+                container.HostContext.Initialize();
+                // Now we can copy it
+                hostContext = container.HostContext;
+            }
+            else
+            {
+                hostContext = HostContext.Create();
+            }
 
-            context.HttpContext = httpContext;
-            return context;
+            hostContext.HttpContext = httpContext;
+            _diagnostics.BeginRequest(httpContext, hostContext);
+
+            return hostContext;
         }
 
         // Execute the request
-        public Task ProcessRequestAsync(Context context)
+        public Task ProcessRequestAsync(HostContext context)
         {
             return _application(context.HttpContext);
         }
 
         // Clean up the request
-        public void DisposeContext(Context context, Exception exception)
+        public void DisposeContext(HostContext context, Exception exception)
         {
             var httpContext = context.HttpContext;
+
             _diagnostics.RequestEnd(httpContext, exception, context);
             _httpContextFactory.Dispose(httpContext);
             _diagnostics.ContextDisposed(context);
+
+            context.Reset();
         }
 
+        // Struct to turn {TContext} uses into faster concrete generic implementations rather than shared generics.
```

Comment:
Just saw this. 👍 Not only does it help future maintainers, it heads off inevitable code review feedback from me 😄